kf/kf_unix.go

```diff
@@ -194,3 +202,176 @@ func VerifyNCreateTCDirs() error {
+				Ifindex: uint32(iface.Index),
+				Handle:  core.BuildHandle(tc.HandleRoot, 0x0000),
+				Parent:  tc.HandleIngress,
+				Info:    0,
+			},
+			Attribute: tc.Attribute{
+				Kind: "clsact",
+			},
+		}
+
+		if err := tcgo.Qdisc().Add(&qdisc); err != nil {
+			log.Info().Msgf("could not assign clsact to %s: %v, its already exists", ifaceName, err)
+		}
+	}
+
+	CollectionRef, err := ebpf.LoadCollection(eBPFProgram.Program.ObjectFile)
+	if err != nil {
+		log.Error().Msgf("loading of tc root %s failed direction %s", eBPFProgram.Program.ObjectFile, direction)
+		return fmt.Errorf("%s: loading of tc root failed", eBPFProgram.Program.ObjectFile)
+	}
+
+	// storing collection reference pointer
+	b.CollectionRef = CollectionRef
+
+	var bpfRootProg *ebpf.Program
+	var bpfRootMap *ebpf.Map
+	var rootArrayMapFileName string
+
+	bpfRootProg = CollectionRef.Programs[eBPFProgram.Program.EntryFunctionName]
+	bpfRootMap = CollectionRef.Maps[strings.Split(eBPFProgram.Program.MapName, "/")[2]]
+	rootArrayMapFileName = filepath.Join(b.hostConfig.BpfMapDefaultPath, eBPFProgram.Program.MapName)
+
+	// Pinning root program
+	if err := bpfRootMap.Pin(rootArrayMapFileName); err != nil {
+		return fmt.Errorf("%s failed to pin the map", rootArrayMapFileName)
+	}
+
+	var parent uint32
+	if direction == models.IngressType {
+		parent = tc.HandleMinIngress
+	} else if direction == models.EgressType {
+		parent = tc.HandleMinEgress
+	}
+
+	progFD := uint32(bpfRootProg.FD())
+	bpfFlag := uint32(0x1)
+	filter := tc.Object{
+		Msg: tc.Msg{
+			Family:  unix.AF_UNSPEC,
+			Ifindex: uint32(iface.Index),
+			Handle:  0,
+			Parent:  core.BuildHandle(tc.HandleRoot, parent),
+			Info:    0x300,
+		},
+		Attribute: tc.Attribute{
+			Kind: "bpf",
+			BPF: &tc.Bpf{
+				FD:    &progFD,
+				Flags: &bpfFlag,
+			},
+		},
+	}
+
+	// Storing Filter handle
+	b.TCFilter = tcgo.Filter()
+
+	// Attaching / Adding as filter
+	if err := b.TCFilter.Add(&filter); err != nil {
+		return fmt.Errorf("could not attach filter for eBPF program: %v", err)
+	}
+
+	return nil
+}
+
+// UnloadTCProgram - Remove TC filters
+func (b *BPF) UnloadTCProgram(ifaceName, direction string) error {
+
+	iface, err := net.InterfaceByName(ifaceName)
+	if err != nil {
+		log.Fatal().Msgf("lookup network iface %q: %s", ifaceName, err)
+	}
+
+	bpfRootProg := b.CollectionRef.Programs[b.Program.EntryFunctionName]
+
+	var parent uint32
+	if direction == models.IngressType {
+		parent = tc.HandleMinIngress
+	} else if direction == models.EgressType {
+		parent = tc.HandleMinEgress
+	}
+
+	tcfilts, err := b.TCFilter.Get(&tc.Msg{
+		Family:  unix.AF_UNSPEC,
+		Ifindex: uint32(iface.Index),
+		Handle:  0x0,
+		Parent:  core.BuildHandle(tc.HandleRoot, tc.HandleMinIngress),
+	})
+
+	if err != nil {
+		log.Warn().Msgf("Could not get filters for interface \"%s\" direction %s ", ifaceName, direction)
+		return fmt.Errorf("could not get filters %v", err)
+	}
+
+	progFD := uint32(bpfRootProg.FD())
+	bpfFlag := uint32(0x1)
```

Comment:
What does 0x1 mean here?

Reply:
Flags defined by the kernel for the BPF filter, value set is direct attach.

Reply:
Please add a comment or constant in the code.